ymmsl/checkpoint.py

```diff
@@ -0,0 +1,77 @@
+"""Definitions for describing checkpoints."""
+
+from typing import List, Optional, Union
+
+import yatiml
+
+
+class CheckpointRange:
+    def __init__(self,
+                 step: Union[float, int],
+                 start: Optional[Union[float, int]] = None,
+                 stop: Optional[Union[float, int]] = None) -> None:
+        if step <= 0:
+            raise ValueError(f"Invalid step {step} in checkpoint range:"
+                             " must be larger than 0")
+        if start is not None and stop is not None and start > stop:
+            raise ValueError(f"Invalid start {start} and stop {stop} in"
+                             "checkpoint range: stop cannot be smaller than"
+                             " start")
+        self.start = start
+        self.stop = stop
+        self.step = step
+
+    @classmethod
+    def _yatiml_sweeten(cls, node: yatiml.Node) -> None:
+        if node.get_attribute("start").is_scalar(type(None)):
+            node.remove_attribute("start")
+
+        if node.get_attribute("stop").is_scalar(type(None)):
+            node.remove_attribute("stop")
+
+
+class CheckpointRules:
+    def __init__(self,
+                 every: Optional[Union[float, int]] = None,
+                 at: Optional[List[Union[float, int]]] = None,
```

Comment:
It would be nice (and also consistent) if `at` took a single float or int, as well as a list of them. That way if you need only one you can just write that. See also `Component.multiplicity`.